pkg/controller/nodelink/nodelink_controller.go

```diff
@@ -0,0 +1,430 @@
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		klog.Errorf("Error getting node: %v", err)
+		return reconcile.Result{}, fmt.Errorf("error getting node: %v", err)
+	}
+
+	machine, err := r.findMachineFromNode(*node)
+	if err != nil {
+		klog.Errorf("Failed to find machine from node %q: %v", node.GetName(), err)
+		return reconcile.Result{}, fmt.Errorf("failed to find machine from node %q: %v", node.GetName(), err)
+	}
+
+	if machine == nil {
+		klog.Warningf("Machine for node %q not found", node.GetName())
+		return reconcile.Result{}, nil
+	}
+
+	modNode := node.DeepCopy()
+	if modNode.Annotations == nil {
+		modNode.Annotations = map[string]string{}
+	}
+	modNode.Annotations[machineAnnotationKey] = fmt.Sprintf("%s/%s", machine.GetNamespace(), machine.GetName())
+
+	if modNode.Labels == nil {
+		modNode.Labels = map[string]string{}
+	}
+
+	for k, v := range machine.Spec.Labels {
+		klog.V(4).Infof("Copying label %s = %s", k, v)
+		modNode.Labels[k] = v
+	}
+
+	addTaintsToNode(modNode, machine)
+
+	if !reflect.DeepEqual(node, modNode) {
+		klog.V(3).Infof("Node %q has changed, updating", modNode.GetName())
+		if err := r.client.Update(context.Background(), modNode); err != nil {
+			return reconcile.Result{}, fmt.Errorf("error updating node: %v", err)
+		}
+	}
+
+	// TODO(alberto): bring nodeRef logic here and drop cluster-api node controller
+	return reconcile.Result{}, nil
+}
+
+// nodeRequestFromMachine returns a reconcile.request for the node backed by the received machine
+func (r *ReconcileNodeLink) nodeRequestFromMachine(o handler.MapObject) []reconcile.Request {
+	klog.V(3).Infof("Watched machine event, finding node to reconcile.Request")
+	// get machine
+	machine := &mapiv1beta1.Machine{}
+	if err := r.client.Get(
+		context.Background(),
+		client.ObjectKey{
+			Namespace: o.Meta.GetNamespace(),
+			Name:      o.Meta.GetName(),
+		},
+		machine,
+	); err != nil {
+		klog.Errorf("No-op: Unable to retrieve machine %s/%s from store: %v", o.Meta.GetNamespace(), o.Meta.GetName(), err)
+		return []reconcile.Request{}
+	}
+
+	if machine.DeletionTimestamp != nil {
+		klog.V(3).Infof("No-op: Machine %q has a deletion timestamp", o.Meta.GetName())
+		return []reconcile.Request{}
+	}
+
+	// find node
+	node, err := r.findNodeFromMachine(*machine)
+	if err != nil {
+		klog.Errorf("No-op: Failed to find node for machine %q: %v", machine.GetName(), err)
+		return []reconcile.Request{}
+	}
+	if node != nil {
+		return []reconcile.Request{
+			{
+				NamespacedName: client.ObjectKey{
+					Namespace: node.GetNamespace(),
+					Name:      node.GetName(),
+				},
+			},
+		}
+	}
+
+	klog.V(3).Infof("No-op: Node for machine %q not found", machine.GetName())
+	return []reconcile.Request{}
+}
+
+// findNodeFromMachine tries to a node from by providerIP and fallback to find by IP
+func (r *ReconcileNodeLink) findNodeFromMachine(machine mapiv1beta1.Machine) (*corev1.Node, error) {
```

Comment:
Any reason not to pass a pointer here? `machine` is not mutated.

Reply:
And the same comment goes to the other find() methods.

Reply:
why would you pass a pointer? that'd mean the machine is mutable within the function, why?

Reply:
Asking because it is not mutated and would prevent the shallow copy.

Reply:
The shallow copy doesn't prevent other pointer values being mutated: 

```go
package main

import (
	"fmt"
)

type A struct {
	b *B
}

type B struct {
	name string
}

func bar(a A) {
	a.b.name = "oops"
}

func main() {
	a := A{&B{name: "hello world"}}
	fmt.Println(a.b.name)

	bar(a)

	if a.b.name != "hello world" {
		panic(a.b.name)
	}
}
```

Reply:
done